mutex.go

```diff
@@ -0,0 +1,35 @@
+/*
+This code was taken from the same implementation in a branch from Consul and then
+had the package updated and the mutex type unexported.
+*/
+package autopilot
+
+import (
+	"context"
+
+	"golang.org/x/sync/semaphore"
+)
+
+type mutex semaphore.Weighted
```

Comment:
Why use `semaphore.Weighted` over `sync.Mutex`?

Reply:
Ah I think I see, it must be because of `TryLock`.

Reply:
sync.Mutex is uncancellable. The semaphore from x/sync can take in a context and allow returning an error instead of waiting indefinitely for the lock. All of this code was taken from some code @dnephin wrote and is being merged into Consul itself.